lib/srv/db/common/report.go

```diff
@@ -0,0 +1,246 @@
+// Copyright 2023 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"context"
+	"net"
+	"runtime"
+	"strings"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/lib/observability/metrics"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+type reporterConfig struct {
+	engine     Engine
+	engineName string
+	clock      clockwork.Clock
+	component  string
+}
+
+func (r *reporterConfig) CheckAndSetDefaults() error {
+	if r.engine == nil {
+		return trace.BadParameter("missing parameter Engine")
+	}
+	if r.engineName == "" {
+		return trace.BadParameter("missing parameter EngineName")
+	}
+	if r.clock == nil {
+		r.clock = clockwork.NewRealClock()
+	}
+	if r.component == "" {
+		r.component = teleport.ComponentDatabase
+	}
+	return nil
+}
+
+type reportingEngine struct {
+	reporterConfig
+}
+
+func init() {
+	_ = metrics.RegisterPrometheusCollectors(prometheusCollectorsEngine...)
+}
+
+func NewReportingEngine(cfg reporterConfig) (Engine, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &reportingEngine{cfg}, nil
+}
+
+func (r *reportingEngine) InitializeConnection(clientConn net.Conn, sessionCtx *Session) error {
+	closer := utils.NewCloserConn(clientConn)
+	trackingClientConn := utils.NewTrackingConn(closer)
+
+	go func() {
+		var txPrev, rxPrev uint64
+
+		update := func() {
+			tx, rx := trackingClientConn.Stat()
+			clientWrittenBytes.WithLabelValues(r.component, r.engineName).Add(float64(tx - txPrev))
+			clientReadBytes.WithLabelValues(r.component, r.engineName).Add(float64(rx - rxPrev))
```

Comment:
The `clientReadBytes`  metrics actually aggregate all reads operation from all connections per engine. 

I wonder  what we are gaining having this low level network metric per clients pool ?

Reply:
We can make it more granular if we'd like. Would you like to see any particular labels added?

On its own, without any additional information, it probably isn't very useful, but on the other hand, many other metrics are like this. It can be very powerful when answering certain questions, though. 

For example: given one client, one Teleport instance, and one database server, what is the maximum throughput that can be achieved? How does this answer change if we increase the latency anywhere? How do different protocols compare? Does compression helps or not?